v2/RNHealthKitCore/HealthKitCore/HealthKitTypes.swift

```diff
@@ -0,0 +1,143 @@
+import Foundation
+import HealthKit
+
+public protocol HealthKitType: RawRepresentable {
+    var type: HKSampleType { get }
```

Comment:
suggestion: Since this protocol is not directly available in objective-c maybe you could add this `HKSampleType` as a primary associated type to the protocol. So you don't need to cast the type when using a concrete type.

I totally understand not going for it if the compiler is giving you grief for using PATs, an alternative could be having a computed property that does the type casting on the concrete types. 
```swift
enum QuantityType {
/// Don't use this directly. Use `quantitySampleType` instead.
public var type: HKSampleType { ... } // I wonder if there's a way to use @available here to help 
public var quantitySampleType: HKQuantityType { type as! HKQuantityType }
```